### PR TITLE
Backend implementation

### PR DESCRIPTION
**Pull request by leogomes**
* Run `mvn clean install && java -jar target/flamegrapher-1.0.0-SNAPSHOT-fat.jar` and you can launch the backend and UI all together. Go to http://localhost:8080/flames;
* With the current state of things, it will fail if there's any Java process that is an OpenJDK one in the machine. Basically, because `jcmd` will list all JVM processes, including OpenJDK ones, but the command `jcmd [PID] JFR.check` in an OpenJDK will result in an exception. Typically, IntelliJ will have embedded processes launched with OpenJDK (at least on Mac). This needs to be fixed, before the PR is merged.

### CODE REVIEW

File: src/main/java/flamegrapher/model/JVM.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package flamegrapher.model;
+
+import static java.util.regex.Pattern.compile;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JVM {
+
+    private static final Pattern VM_VERSION = compile("JDK (?<majorVersion>[0-9])\\.",
```
Comment (nbogojevic) [inline src/main/java/flamegrapher/model/JVM.java]: Does this work with two digit jdk (i.e. 10)?

File: src/main/java/flamegrapher/backend/JavaFlightRecorder.java [hunk 2]
```diff
@@ -53,51 +51,101 @@ public JavaFlightRecorder(Vertx vertx) {
...
+                return json;
+            });
+            
+        }, handler);
+    }
+
+    private List<String> argumentsForDump(String pid, String recording, String filename, JVM jvm) {
+        List<String> args;
+        // Command changed starting with JDK 9
+        if (jvm.getMajorVersion() == 9) {
```
Comment (nbogojevic) [inline src/main/java/flamegrapher/backend/JavaFlightRecorder.java]: Maket it > 8 as in #6

File: src/test/java/flamegrapher/model/JVMTest.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        JVM vm = JVM.fromVMVersion(output);
+        assertThat(vm.getMajorVersion(), equalTo(8));
+        
+    }
+    
+    @Test
+    public void version9() {
+        String output = "14785:\n" + 
+                "Java HotSpot(TM) 64-Bit Server VM version 9+181\n" + 
+                "JDK 9.0.0";
```
Comment (nbogojevic) [inline src/test/java/flamegrapher/model/JVMTest.java]: Should test against supported versions ;)

File: src/test/java/flamegrapher/model/ProcessesTest.java [hunk 4]
```diff
@@ -9,16 +9,69 @@
...
+    @Test
+    public void checkStateFriendlyName() {
+        assertThat(State.NOT_RECORDING.toString(), equalTo("Not recording"));
+    }
+    
+    @Test
+    public void startStatus() {
+        String output = "73191:\n" + 
+                "Stopped recording 4.\n" + 
+                "NCELRND0380:flamegrapher lgomes$ jcmd 73191 JFR.start\n" +
```
Comment (nbogojevic) [inline src/test/java/flamegrapher/model/ProcessesTest.java]: oops!

File: src/main/java/flamegrapher/backend/JavaFlightRecorder.java [hunk 5]
```diff
@@ -53,51 +51,101 @@ public JavaFlightRecorder(Vertx vertx) {
...
+            args = asList(pid, "JFR.dump", "filename=" + filename, "name=" + recording);
```
Comment (nbogojevic) [inline src/main/java/flamegrapher/backend/JavaFlightRecorder.java]: On JDK 10 it seems to require full name, not just number. Will this recording be "1" or "Recording 1"?

File: src/main/java/flamegrapher/backend/JavaFlightRecorder.java [hunk 6]
```diff
@@ -106,8 +154,27 @@ private String filename(String pid, String recording) {
...
-        jcmd(asList(pid, "JFR.stop", "recording=" + recording), handler);
+    public void stop(String pid, String recording, Future<Item> handler) {
+        
+        // First we need to check the JVM version
+        Future<JVM> jvmFuture = Future.future();
+        jcmd(asList(pid, "VM.version"), jvmFuture, JVM::fromVMVersion);
+
+        // Once we got the JVM version, we can execute the stop
+        jvmFuture.compose(jvm -> {
+            String recordingParameter = (jvm.getMajorVersion() == 9 ? "name=" : "recording=");
```
Comment (nbogojevic) [inline src/main/java/flamegrapher/backend/JavaFlightRecorder.java]: `>= 9`

File: src/main/java/flamegrapher/model/Processes.java [hunk 7]
```diff
@@ -3,30 +3,28 @@
...
+        return items;
     }
 
     public static Processes fromString(String string) {
         Processes p = new Processes();
         String[] lines = string.split("\n");
         for(String line : lines) {
             String[] parts = line.split(" ", 2);
-            p.pidName.put(parts[0], parts[1]);
+            String pid = parts[0];
```
Comment (nbogojevic) [inline src/main/java/flamegrapher/model/Processes.java]: how about using .split(System.lineSeparator()) in line 20 to make sure it works on Windows

File: src/test/java/flamegrapher/backend/JfrParserTest.java [hunk 8]
```diff
@@ -12,6 +13,7 @@
 
 public class JfrParserTest {
 
+    @Ignore
```
Comment (nbogojevic) [inline src/test/java/flamegrapher/backend/JfrParserTest.java:16]: Can we remove it completely or add a .jfr as an example in test resources

**Review by nbogojevic — COMMENTED** [review]
Few comments

File: src/main/java/flamegrapher/backend/JavaFlightRecorder.java (hunk 5 above)
Comment (leogomes) [reply src/main/java/flamegrapher/backend/JavaFlightRecorder.java]: name=1, gave it a try with both 10 and 9 :)

It's misleading though :)

File: src/main/java/flamegrapher/backend/JavaFlightRecorder.java (hunk 6 above)
Comment (leogomes) [reply src/main/java/flamegrapher/backend/JavaFlightRecorder.java]: Should be refactored indeed.

File: src/test/java/flamegrapher/backend/JfrParserTest.java (hunk 8 above)
Comment (leogomes) [reply src/test/java/flamegrapher/backend/JfrParserTest.java:16]: I was planning to do so (add the .jfr), just didn't have time yet.

Comment (leogomes) [reply src/test/java/flamegrapher/backend/JfrParserTest.java:16]: https://github.com/flamegrapher/flamegrapher/issues/8 
Added to handle this.

**Comment by leogomes** [comment]
@nbogojevic Fixed most of your comments. I'm done with this one, unless you have other comments :) This is relatively usable already (UI + backend in the fat jar).

**Pull request merged**